File: lib/security_groups.py
from constructs import Construct
from cdktf_cdktf_provider_aws.security_group import SecurityGroup
from cdktf_cdktf_provider_aws.security_group_rule import SecurityGroupRule


class SecurityGroups(Construct):
    def __init__(self, scope: Construct, id: str, *,  # pylint: disable=redefined-builtin,too-many-positional-arguments
                 environment_suffix: str, vpc_id: str, vpc_cidr: str):
        super().__init__(scope, id)

        # EKS Cluster Security Group
        self.cluster_sg = SecurityGroup(self, "eks_cluster_sg",
            name=f"eks-cluster-sg-{environment_suffix}",
            description="Security group for EKS cluster control plane",
            vpc_id=vpc_id,
            tags={
                "Name": f"eks-cluster-sg-{environment_suffix}",
                "Environment": environment_suffix
            }
        )

        # Allow HTTPS ingress from VPC CIDR
        SecurityGroupRule(self, "cluster_ingress_443",
            type="ingress",
            from_port=443,
            to_port=443,
            protocol="tcp",
            cidr_blocks=[vpc_cidr],
            security_group_id=self.cluster_sg.id,
            description="Allow HTTPS from VPC"
        )

        # Allow all egress
        SecurityGroupRule(self, "cluster_egress_all",
            type="egress",
            from_port=0,
            to_port=0,
            protocol="-1",
            cidr_blocks=["0.0.0.0/0"],
            security_group_id=self.cluster_sg.id,
            description="Allow all outbound traffic"
        )

    @property
    def cluster_security_group_id(self) -> str:
        return self.cluster_sg.id

File: archive/pulumi-ts/Pr7159/lib/components/security-groups.ts
import * as pulumi from '@pulumi/pulumi';
import * as aws from '@pulumi/aws';

export interface SecurityGroupsArgs {
  vpcId: pulumi.Input<string>;
  environmentSuffix: string;
  tags: { [key: string]: string };
}

export class SecurityGroupsComponent extends pulumi.ComponentResource {
  public readonly albSecurityGroup: aws.ec2.SecurityGroup;
  public readonly ecsSecurityGroup: aws.ec2.SecurityGroup;
  public readonly rdsSecurityGroup: aws.ec2.SecurityGroup;

  constructor(
    name: string,
    args: SecurityGroupsArgs,
    opts?: pulumi.ComponentResourceOptions
  ) {
    super('custom:security:SecurityGroupsComponent', name, {}, opts);

    const defaultResourceOptions: pulumi.ResourceOptions = { parent: this };

    // ALB Security Group
    this.albSecurityGroup = new aws.ec2.SecurityGroup(
      `alb-sg-${args.environmentSuffix}`,
      {
        vpcId: args.vpcId,
        description: 'Security group for Application Load Balancer',
        ingress: [
          {
            protocol: 'tcp',
            fromPort: 80,
            toPort: 80,
            cidrBlocks: ['0.0.0.0/0'],
            description: 'Allow HTTP from anywhere',
          },
          {
            protocol: 'tcp',
            fromPort: 443,
            toPort: 443,
            cidrBlocks: ['0.0.0.0/0'],
            description: 'Allow HTTPS from anywhere',
          },
        ],
        egress: [
          {
            protocol: '-1',
            fromPort: 0,
            toPort: 0,
            cidrBlocks: ['0.0.0.0/0'],
            description: 'Allow all outbound traffic',
          },
        ],
        tags: {
          ...args.tags,
          Name: `alb-sg-${args.environmentSuffix}`,
        },
      },
      defaultResourceOptions
    );

    // ECS Security Group
    this.ecsSecurityGroup = new aws.ec2.SecurityGroup(
      `ecs-sg-${args.environmentSuffix}`,
      {
        vpcId: args.vpcId,
        description: 'Security group for ECS tasks',
        ingress: [
          {
            protocol: 'tcp',
            fromPort: 8080,
            toPort: 8080,
            securityGroups: [this.albSecurityGroup.id],
            description: 'Allow traffic from ALB',
          },
        ],
        egress: [
          {
            protocol: '-1',
            fromPort: 0,
            toPort: 0,
            cidrBlocks: ['0.0.0.0/0'],
            description: 'Allow all outbound traffic',
          },
        ],
        tags: {
          ...args.tags,
          Name: `ecs-sg-${args.environmentSuffix}`,
        },
      },
      defaultResourceOptions
    );

    // RDS Security Group
    this.rdsSecurityGroup = new aws.ec2.SecurityGroup(
      `rds-sg-${args.environmentSuffix}`,
      {
        vpcId: args.vpcId,
        description: 'Security group for RDS database',
        ingress: [
          {
            protocol: 'tcp',
            fromPort: 3306,
            toPort: 3306,
            securityGroups: [this.ecsSecurityGroup.id],
            description: 'Allow traffic from ECS tasks',
          },
        ],
        egress: [
          {
            protocol: '-1',
            fromPort: 0,
            toPort: 0,
            cidrBlocks: ['0.0.0.0/0'],
            description: 'Allow all outbound traffic',
          },
        ],
        tags: {
          ...args.tags,
          Name: `rds-sg-${args.environmentSuffix}`,
        },
      },
      defaultResourceOptions
    );

    this.registerOutputs({
      albSecurityGroupId: this.albSecurityGroup.id,
      ecsSecurityGroupId: this.ecsSecurityGroup.id,
      rdsSecurityGroupId: this.rdsSecurityGroup.id,
    });
  }
}
